## Ana Ganfornina

In [1]:
import pandas as pd

In [2]:
data =  pd.read_csv(r"./dataSet/kiva_loans.csv")

In [3]:
data

,id,funded_amount,loan_amount,activity,sector,use,country_code,country,region,currency,partner_id,posted_time,disbursed_time,funded_time,term_in_months,lender_count,tags,borrower_genders,repayment_interval,date
0,653051,300.0,300.0,Fruits & Vegetables,Food,"To buy seasonal, fresh fruits to sell.",PK,Pakistan,Lahore,PKR,247.0,2014-01-01 06:12:39+00:00,2013-12-17 08:00:00+00:00,2014-01-02 10:06:32+00:00,12.0,12,NaN,female,irregular,2014-01-01
1,653053,575.0,575.0,Rickshaw,Transportation,to repair and maintain the auto rickshaw used ...,PK,Pakistan,Lahore,PKR,247.0,2014-01-01 06:51:08+00:00,2013-12-17 08:00:00+00:00,2014-01-02 09:17:23+00:00,11.0,14,NaN,"female, female",irregular,2014-01-01
2,653068,150.0,150.0,Transportation,Transportation,To repair their old cycle-van and buy another ...,IN,India,Maynaguri,INR,334.0,2014-01-01 09:58:07+00:00,2013-12-17 08:00:00+00:00,2014-01-01 16:01:36+00:00,43.0,6,"user_favorite, user_favorite",female,bullet,2014-01-01
3,653063,200.0,200.0,Embroidery,Arts,to purchase an embroidery machine and a variet...,PK,Pakistan,Lahore,PKR,247.0,2014-01-01 08:03:11+00:00,2013-12-24 08:00:00+00:00,2014-01-01 13:00:00+00:00,11.0,8,NaN,female,irregular,2014-01-01
4,653084,400.0,400.0,Milk Sales,Food,to purchase one buffalo.,PK,Pakistan,Abdul Hakeem,PKR,245.0,2014-01-01 11:53:19+00:00,2013-12-17 08:00:00+00:00,2014-01-01 19:18:51+00:00,14.0,16,NaN,female,monthly,2014-01-01
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
671200,1340323,0.0,25.0,Livestock,Agriculture,"[True, u'para compara: cemento, arenya y ladri...",PY,Paraguay,Concepción,USD,58.0,2017-07-25 16:55:34+00:00,2017-07-25 07:00:00+00:00,NaN,13.0,0,NaN,female,monthly,2017-07-25
671201,1340316,25.0,25.0,Livestock,Agriculture,"[True, u'to start a turducken farm.'] - this l...",KE,Kenya,NaN,KES,138.0,2017-07-25 06:14:08+00:00,2017-07-24 07:00:00+00:00,2017-07-26 02:09:43+00:00,13.0,1,NaN,female,monthly,2017-07-25
671202,1340334,0.0,25.0,Games,Entertainment,NaN,KE,Kenya,NaN,KES,138.0,2017-07-26 00:02:07+00:00,2017-07-25 07:00:00+00:00,NaN,13.0,0,NaN,NaN,monthly,2017-07-26
671203,1340338,0.0,25.0,Livestock,Agriculture,"[True, u'to start a turducken farm.'] - this l...",KE,Kenya,NaN,KES,138.0,2017-07-26 06:12:55+00:00,2017-07-25 07:00:00+00:00,NaN,13.0,0,NaN,female,monthly,2017-07-26


In [4]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 671205 entries, 0 to 671204
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  671205 non-null  int64  
 1   funded_amount       671205 non-null  float64
 2   loan_amount         671205 non-null  float64
 3   activity            671205 non-null  str    
 4   sector              671205 non-null  str    
 5   use                 666973 non-null  str    
 6   country_code        671197 non-null  str    
 7   country             671205 non-null  str    
 8   region              614405 non-null  str    
 9   currency            671205 non-null  str    
 10  partner_id          657698 non-null  float64
 11  posted_time         671205 non-null  str    
 12  disbursed_time      668809 non-null  str    
 13  funded_time         622874 non-null  str    
 14  term_in_months      671205 non-null  float64
 15  lender_count        671205 non-null  int64  


In [5]:
data.shape

(671205, 20)

In [ ]:
data.describe() # parecido al estatistics de excel


,id,funded_amount,loan_amount,partner_id,term_in_months,lender_count
count,6.712050e+05,671205.000000,671205.000000,657698.000000,671205.000000,671205.000000
mean,9.932486e+05,785.995061,842.397107,178.199616,13.739022,20.590922
std,1.966113e+05,1130.398941,1198.660073,94.247581,8.598919,28.459551
min,6.530470e+05,0.000000,25.000000,9.000000,1.000000,0.000000
25%,8.230720e+05,250.000000,275.000000,126.000000,8.000000,7.000000
50%,9.927800e+05,450.000000,500.000000,145.000000,13.000000,13.000000
75%,1.163653e+06,900.000000,1000.000000,204.000000,14.000000,24.000000
max,1.340339e+06,100000.000000,100000.000000,536.000000,158.000000,2986.000000


## Limpieza de datos

### Limpieza de fechas

- Pasar de tipo str a datetime las fechas

In [49]:
data_type = data.loc[:,['posted_time', 'disbursed_time', 'funded_time', 'date']]
data_type.info()

<class 'pandas.DataFrame'>
Index: 622633 entries, 0 to 671201
Data columns (total 4 columns):
 #   Column          Non-Null Count   Dtype              
---  ------          --------------   -----              
 0   posted_time     622633 non-null  datetime64[us, UTC]
 1   disbursed_time  622633 non-null  datetime64[us, UTC]
 2   funded_time     622633 non-null  datetime64[us, UTC]
 3   date            622633 non-null  datetime64[us]     
dtypes: datetime64[us, UTC](3), datetime64[us](1)
memory usage: 23.8 MB


In [50]:
colum_date = ['posted_time', 'disbursed_time', 'funded_time', 'date']
for column in colum_date:
    data[column] = pd.to_datetime(data[column]) #Por cada columna cambia a tipo fecha

data_type = data.loc[:,['posted_time', 'disbursed_time', 'funded_time', 'date']]
data_type.info()

<class 'pandas.DataFrame'>
Index: 622633 entries, 0 to 671201
Data columns (total 4 columns):
 #   Column          Non-Null Count   Dtype              
---  ------          --------------   -----              
 0   posted_time     622633 non-null  datetime64[us, UTC]
 1   disbursed_time  622633 non-null  datetime64[us, UTC]
 2   funded_time     622633 non-null  datetime64[us, UTC]
 3   date            622633 non-null  datetime64[us]     
dtypes: datetime64[us, UTC](3), datetime64[us](1)
memory usage: 23.8 MB


### Comprobación de nulos

In [51]:
data.isnull()

,id,activity,sector,use,country,region,currency,partner_id,posted_time,disbursed_time,funded_time,term_in_months,lender_count,borrower_genders,repayment_interval,date,funded_amount_usd,loan_amount_usd
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
671166,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
671170,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
671177,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
671186,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [52]:
data.isnull().mean() * 100

id                    0.000000
activity              0.000000
sector                0.000000
use                   0.000000
country               0.000000
region                0.000000
currency              0.000000
partner_id            0.000000
posted_time           0.000000
disbursed_time        0.000000
funded_time           0.000000
term_in_months        0.000000
lender_count          0.000000
borrower_genders      0.000000
repayment_interval    0.000000
date                  0.000000
funded_amount_usd     0.000321
loan_amount_usd       0.000321
dtype: float64

In [54]:
data = data.drop(columns=['tags']) #Eliminamos tags debido a que tiene un 25% de nulos
data.info()

KeyError: "['tags'] not found in axis"

- Sustitución de nulos por "No especificado" en caso de string y por la media numerica de la columna en caso de número

In [ ]:

#Seleccionar los nombres de las columnas que son texto
columns_str = data.select_dtypes(str).columns
print(columns_str)

for column in columns_str:
    data[column] = data[column].fillna("No especificado")

# Seleccionamos las que son int o float
columns_num = data.select_dtypes('number').columns

for column in columns_num:
    # Calculamos la mediana de la columna
    mediana = data[column].median()
    # Rellenamos los nulos de esa columna con su mediana
    data[column] = data[column].fillna(mediana)

# Verificamos que ya no quedan nulos
print(data.isnull().sum())




Index(['activity', 'sector', 'use', 'country_code', 'country', 'region',
       'currency', 'borrower_genders', 'repayment_interval'],
      dtype='str')
id                        0
funded_amount             0
loan_amount               0
activity                  0
sector                    0
use                       0
country_code              0
country                   0
region                    0
currency                  0
partner_id                0
posted_time               0
disbursed_time         2396
funded_time           48331
term_in_months            0
lender_count              0
borrower_genders          0
repayment_interval        0
date                      0
dtype: int64


In [ ]:
data

,id,funded_amount,loan_amount,activity,sector,use,country_code,country,region,currency,partner_id,posted_time,disbursed_time,funded_time,term_in_months,lender_count,borrower_genders,repayment_interval,date
0,653051,300.0,300.0,Fruits & Vegetables,Food,"To buy seasonal, fresh fruits to sell.",PK,Pakistan,Lahore,PKR,247.0,2014-01-01 06:12:39+00:00,2013-12-17 08:00:00+00:00,2014-01-02 10:06:32+00:00,12.0,12,female,irregular,2014-01-01
1,653053,575.0,575.0,Rickshaw,Transportation,to repair and maintain the auto rickshaw used ...,PK,Pakistan,Lahore,PKR,247.0,2014-01-01 06:51:08+00:00,2013-12-17 08:00:00+00:00,2014-01-02 09:17:23+00:00,11.0,14,"female, female",irregular,2014-01-01
2,653068,150.0,150.0,Transportation,Transportation,To repair their old cycle-van and buy another ...,IN,India,Maynaguri,INR,334.0,2014-01-01 09:58:07+00:00,2013-12-17 08:00:00+00:00,2014-01-01 16:01:36+00:00,43.0,6,female,bullet,2014-01-01
3,653063,200.0,200.0,Embroidery,Arts,to purchase an embroidery machine and a variet...,PK,Pakistan,Lahore,PKR,247.0,2014-01-01 08:03:11+00:00,2013-12-24 08:00:00+00:00,2014-01-01 13:00:00+00:00,11.0,8,female,irregular,2014-01-01
4,653084,400.0,400.0,Milk Sales,Food,to purchase one buffalo.,PK,Pakistan,Abdul Hakeem,PKR,245.0,2014-01-01 11:53:19+00:00,2013-12-17 08:00:00+00:00,2014-01-01 19:18:51+00:00,14.0,16,female,monthly,2014-01-01
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
671200,1340323,0.0,25.0,Livestock,Agriculture,"[True, u'para compara: cemento, arenya y ladri...",PY,Paraguay,Concepción,USD,58.0,2017-07-25 16:55:34+00:00,2017-07-25 07:00:00+00:00,NaT,13.0,0,female,monthly,2017-07-25
671201,1340316,25.0,25.0,Livestock,Agriculture,"[True, u'to start a turducken farm.'] - this l...",KE,Kenya,No especificado,KES,138.0,2017-07-25 06:14:08+00:00,2017-07-24 07:00:00+00:00,2017-07-26 02:09:43+00:00,13.0,1,female,monthly,2017-07-25
671202,1340334,0.0,25.0,Games,Entertainment,No especificado,KE,Kenya,No especificado,KES,138.0,2017-07-26 00:02:07+00:00,2017-07-25 07:00:00+00:00,NaT,13.0,0,No especificado,monthly,2017-07-26
671203,1340338,0.0,25.0,Livestock,Agriculture,"[True, u'to start a turducken farm.'] - this l...",KE,Kenya,No especificado,KES,138.0,2017-07-26 06:12:55+00:00,2017-07-25 07:00:00+00:00,NaT,13.0,0,female,monthly,2017-07-26


Debido a que los  2396 valores nulos de disbursed_time y los  48331 de funded_time representan solo el 0.36% y 7.2%  del dataSet respectivamente, serán eliminados     

In [ ]:
data = data.dropna() #borra las filas que tengan valores faltantes
data.info()

<class 'pandas.DataFrame'>
Index: 622633 entries, 0 to 671201
Data columns (total 19 columns):
 #   Column              Non-Null Count   Dtype              
---  ------              --------------   -----              
 0   id                  622633 non-null  int64              
 1   funded_amount       622633 non-null  float64            
 2   loan_amount         622633 non-null  float64            
 3   activity            622633 non-null  str                
 4   sector              622633 non-null  str                
 5   use                 622633 non-null  str                
 6   country_code        622633 non-null  str                
 7   country             622633 non-null  str                
 8   region              622633 non-null  str                
 9   currency            622633 non-null  str                
 10  partner_id          622633 non-null  float64            
 11  posted_time         622633 non-null  datetime64[us, UTC]
 12  disbursed_time      622633 non-n

In [ ]:
print(data.isnull().sum())

id                    0
funded_amount         0
loan_amount           0
activity              0
sector                0
use                   0
country_code          0
country               0
region                0
currency              0
partner_id            0
posted_time           0
disbursed_time        0
funded_time           0
term_in_months        0
lender_count          0
borrower_genders      0
repayment_interval    0
date                  0
dtype: int64


### Eliminación de duplicados

In [ ]:
total_duplicados = data.duplicated().sum() #cuenta el número de duplicados
print(f"Tienes {total_duplicados} filas duplicadas.") #como no hay ninguno, no hacemos nada.

Tienes 0 filas duplicadas.


### Normalización de textos

In [ ]:
for column in data.columns:
    if data[column].dtype == 'str':
        data[column] = data[column].str.strip().str.capitalize() #Eliminamos espacios en blanco y ponemos la primera letra en mayusculas



### Normalización de monedas

- Eliminación de country code debido a redundancia con la columna de country y conversión de currency a dolares para poder eliminar la misma columna de country

In [ ]:
data = data.drop(columns=['country_code']) #Eliminación de columna contry_code

In [ ]:
import requests

# Obtenemos tasas de cambio
response = requests.get('https://api.exchangerate-api.com/v4/latest/USD')
rates_usd_base = response.json()['rates']

#Creamos un diccionario vacio al cual decimos que recorra el de rates_usd_base para invertir los valores
rates_to_usd = {}
for currency, rate in rates_usd_base.items():
    rates_to_usd[currency] = 1 / rate

data['currency'] = data['currency'].str.upper() #normalizamos a mayusculas para que coincidan con los datos de conversion de la API

def get_rate(currency_code):
    return rates_to_usd.get(currency_code, None) # En caso de no encontrar la moneda devolveremos None


#Creamos dos nuevas columnas para meter estos nuevos datos
data['funded_amount_usd'] = data['funded_amount'] * data['currency'].apply(get_rate)
data['loan_amount_usd'] = data['loan_amount'] * data['currency'].apply(get_rate)


In [ ]:
#Eliminar  funded_amount y loan_amount puesto que ya no son necesarias y podemos calcularas en base a funded_amount_usd y la tasa de conversión
data = data.drop(columns=['funded_amount','loan_amount'])

Después de releer la documentación de Kiva nos percatamos de que las cifras de funded_amount y loan_amout ya estaban en USD por lo que el código anterior no tendría mucho sentido, pero deja abierta una via de investigación para evaluar la relevancia de los prestamos según el pais.